File: Algorithms/C++/Tree/113_Path_Sum_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> paths;

        if (root == NULL) {
            return paths;
        }

        vector<int> path;
        helper(paths, path, root, sum);

        return paths;
    }
private:
    void helper(vector<vector<int>>& paths, vector<int>& path, TreeNode* root, int sum) {
        if (root == NULL) {
            return;
        }

        path.push_back(root->val);

        if (root->left == NULL && root->right == NULL) {
            if (root->val == sum) {
                paths.push_back(path);
            }
        } else {
            helper(paths, path, root->left, sum - root->val);
            helper(paths, path, root->right, sum - root->val);
        }

        path.pop_back();
    }
};

File: Algorithms/Swift/Tree/113_Path_Sum_II.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func pathSum(root: TreeNode?, _ sum: Int) -> [[Int]] {
        var paths = [[Int]]()
        
        if root == nil {
            return paths
        }
        
        var path = [Int]()
        helper(root, sum: sum, path: &path, paths: &paths)
        
        return paths
    }
    
    private func helper(root: TreeNode?, sum: Int, inout path: [Int], inout paths: [[Int]]) {
        if root == nil {
            return
        }
        
        path.append(root!.val)
        
        if root!.left == nil && root!.right == nil {
            if root!.val == sum {
                paths.append(path)
            }
            path.popLast()
            return
        }
        
        let value = root!.val
        helper(root!.left, sum: sum - value, path: &path, paths: &paths)
        helper(root!.right, sum: sum - value, path: &path, paths: &paths)
        
        path.popLast()
    }
}
